x/committee/types/permissions.go

```diff
@@ -122,14 +123,154 @@ func (changes AllowedParamsChanges) FilterByParamChange(paramChange paramspropos
 	return filtered
 }
 
-// Allows returns true if the given proposal param change is allowed by the AllowedParamsChange rules.
-func (allowed AllowedParamsChange) Allows(paramsChange paramsproposal.ParamChange) bool {
+// SubparamChanges is a map of sub param change keys and its values.
+type SubparamChanges map[string]interface{}
+
+// MultiSubparamChanges is a slice of SubparamChanges.
+type MultiSubparamChanges []SubparamChanges
+
+func (allowed AllowedParamsChange) allowsMultiParamsChange(currentRecords MultiSubparamChanges, incomingRecords MultiSubparamChanges) bool {
+	// do not allow new records from being added or removed for multi-subparam changes.
+	if len(currentRecords) != len(incomingRecords) {
+		return false
+	}
+
+	for _, current := range currentRecords {
+		// find the incoming record and the requirements for each current record
+		var req *SubparamRequirement
+		var incoming *SubparamChanges
+		for _, v := range allowed.MultiSubparamsRequirements {
+			if current[v.Key] == v.Val {
+				req = &v
+				break
+			}
+		}
+
+		// all records should have a requirement, otherwise the change is not allowed
+		if req == nil {
+			return false
+		}
+
+		for _, v := range incomingRecords {
+			if v[req.Key] == req.Val {
+				incoming = &v
+				break
+			}
+		}
+
+		// disallow the change if no incoming record found for current record.
+		if incoming == nil {
+			return false
+		}
+
+		// check incoming changes are allowed
+		allowed := validateParamChangesAreAllowed(current, *incoming, req.AllowedSubparamAttrChanges)
+
+		if !allowed {
+			return false
+		}
+	}
+
+	return true
+}
+
+func validateParamChangesAreAllowed(current SubparamChanges, incoming SubparamChanges, allowList []string) bool {
+	// make sure we are not adding or removing any new attributes
+	if len(current) != len(incoming) {
+		return false
+	}
+
+	for k, v := range current {
+		isAllowed := false
+
+		// check if the param attr key is in the allow list
+		for _, allowedKey := range allowList {
+			if k == allowedKey {
+				isAllowed = true
+				break
+			}
+		}
+
+		// if not allowed, incoming value needs to be the same, or it is rejected
+		if !isAllowed {
+			// since we cannot compare maps directly, we need to convert them to json first.
+			// this should be fine since the data we are marshalling here should always be pretty small.
+			if reflect.TypeOf(v).Kind() == reflect.Map {
+				data, err := json.Marshal(v)
+				if err != nil {
+					return false
+				}
+				data2, err := json.Marshal(incoming[k])
+				if err != nil {
+					return false
+				}
+				if string(data) != string(data2) {
+					return false
+				}
+			} else if v != incoming[k] {
+				return false
+			}
```

Comment:
Couple of thoughts:
- if json marshalling the map isn't deterministic, the string comparison could fail incorrectly. Looks like the current implementation is deterministic, but doesn't guarantee it will always be. https://stackoverflow.com/questions/44755089/does-serialized-content-strictly-follow-the-order-in-definition-use-encoding-json-package
- pricefeed Markets contains a field `Oracles` which is a slice of addresses. This might cause the `else if v != incoming[k]` comparison to fail as slices can't be compared.

Reply:
:+1: updated to use `reflect.DeepEqual` after our discussion. https://github.com/Kava-Labs/kava/pull/1112/commits/327afa4c87e90b0cf03dc3a804e2ca590b5b2b26